src/async-wrap.cc

```diff
@@ -233,26 +265,38 @@ Local<Value> AsyncWrap::MakeCallback(const Local<Function> cb,
     }
   }
 
-  Environment::TickInfo* tick_info = env()->tick_info();
-
   if (callback_scope.in_makecallback()) {
     return ret;
   }
 
+  Environment::TickInfo* tick_info = env()->tick_info();
+
   if (tick_info->length() == 0) {
     env()->isolate()->RunMicrotasks();
   }
 
+  bool next_tick_threw = false;
+
   if (tick_info->length() == 0) {
     tick_info->set_index(0);
-    return ret;
+  } else {
+    next_tick_threw =
+        env()->tick_callback_function()->Call(process, 0, nullptr).IsEmpty();
   }
 
-  if (env()->tick_callback_function()->Call(process, 0, nullptr).IsEmpty()) {
-    return Local<Value>();
+  if (ran_init_callback() && !final_fn.IsEmpty()) {
+    Local<Value> did_throw = Boolean::New(env()->isolate(), next_tick_threw);
+    Local<Value> vals[] = { uid, did_throw };
+    TryCatch try_catch(env()->isolate());
+    Local<Value> ar = final_fn->Call(context, ARRAY_SIZE(vals), vals);
+    if (ar.IsEmpty()) {
+      ClearFatalExceptionHandlers(env()->isolate());
+      FatalException(env()->isolate(), try_catch);
+      return Local<Value>();
+    }
   }
 
-  return ret;
+  return next_tick_threw ? Local<Value>() : ret;
```

Comment:
If `ret.IsEmpty() == true` then `MakeCallback` will have returned early. Just after running the `post` callback (if there is one).